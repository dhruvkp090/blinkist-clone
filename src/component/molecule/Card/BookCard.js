import React from 'react';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import ButtonItem from '../Button/ButtonItem';
import Typography from '@material-ui/core/Typography';
import AddIcon from '@material-ui/icons/Add';
import './Card.css';

const BookCard = (props) => {
    return (
        <Card>
            <CardMedia
                component="img"
                alt={props.bookTitle}
                image= {props.img ? 
                            props.img : 
                            "https://images.blinkist.com/images/books/534d4cb46633610007320000/1_1/640.jpg"}
                title={props.bookTitle}
            />
            <CardContent>
                <Typography gutterBottom variant="h6">
                    {props.bookTitle}
                </Typography>
                <Typography className="author" gutterBottom variant="body1" color="secondary">
                    <strong>{props.bookAuthor}</strong>
                </Typography>
                <div className="timeSection">
                    <img 
                        src="https://d17pjsg7x52x9r.cloudfront.net/assets/components/book_card/clock-b0e2e0235fbe1df824d662b2b3b96611e3711bf5b5c7556b8bd3828720f86dbc.svg" 
                        alt="" className="icon" />
                    <Typography gutterBottom variant="body2" className="timeText">{props.bookDuration}-minute read</Typography>
                </div>
            </CardContent>
            <CardActions >
                {props.isInLibrary(props.id) ? 
                    (changeButtonBasedOnReadStatus(props)) :
                    <ButtonItem startIcon= { <AddIcon /> } 
                                onClick={(e) => props.addToLibrary(e, props.id)} 
                                size="medium" color="primary" 
                                variant="contained" 
                                fullWidth={true}>
                        <Typography variant="body1"><strong>Add to library</strong></Typography>
                    </ButtonItem>
                }
            </CardActions>
    </Card>
    );
};

BookCard.defaultProps = {
    bookTitle: "Animal Farm",
    bookAuthor: "Gorge Orwell",
    bookDuration: "20"
};

export default BookCard;

function changeButtonBasedOnReadStatus(props) {
    return props.isFinished ?
        <ButtonItem 
            onClick={(e) => props.onClick(e, props.id)} 
            size="medium" color="secondary" 
            variant="contained" 
            fullWidth={true}>
            <Typography variant="body1"><strong>Read Again</strong></Typography>
        </ButtonItem> :
        <ButtonItem 
            onClick={(e) => props.onClick(e, props.id)} 
            size="medium" 
            color="primary" 
            variant="contained" 
            fullWidth={true}>
            <Typography variant="body1"><strong>Mark as Finished</strong></Typography>
        </ButtonItem>;
}

