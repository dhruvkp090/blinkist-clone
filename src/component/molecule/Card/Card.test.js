import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react';
import BookCard from './BookCard';
import '@testing-library/jest-dom';

const isInLibraryMock = jest.fn();
const onClickMock = jest.fn();

it('Default card in library', () => {
    const { getByText } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} />);

    getByText("Animal Farm");
    getByText("Gorge Orwell");
    getByText("20-minute read");
    getByText("Mark as Finished");
});

it('Default card not in library', () => {
    render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(false)} />);

    screen.getByText("Animal Farm");
    screen.getByText("Gorge Orwell");
    screen.getByText("20-minute read");
    screen.getByText("Add to library");
});

it('Default card in library finished reading', () => {
    const { getByText } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={true} />);

    getByText("Animal Farm");
    getByText("Gorge Orwell");
    getByText("20-minute read");
    getByText("Read Again");
    expect(screen.queryByText("Add to library")).toBeNull();
    expect(screen.queryByText("Mark as Finished")).toBeNull();
});

it('Snapshot with default image and finished book', () => {
    const { asFragment } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={true} />);

    expect(asFragment()).toMatchSnapshot();
});

it('Snapshot with passed image and unfinished book', () => {
    const { asFragment } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={false} 
        img= "https://images.blinkist.com/images/books/608aa9b16cee070007228a70/1_1/640.jpg"
    />);

    expect(asFragment()).toMatchSnapshot();
});

it('Fire event button click changes status from Mark as finished to read again', () => {
    const { rerender } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={false} onClick={onClickMock} />);
    screen.getByText("Mark as Finished");
    fireEvent.click(screen.getByText('Mark as Finished'));
    rerender(
        <BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={true} onClick={onClickMock} />
    );
    screen.getByText("Read Again");
    expect(onClickMock).toHaveBeenCalledTimes(1);
});

it('Fire event button click changes status from Add to library to Mark as finished', () => {
    const { rerender } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(false)} isFinished={false} addToLibrary={onClickMock} />);
    screen.getByText("Add to library");
    fireEvent.click(screen.getByText('Add to library'));
    rerender(
        <BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={false} onClick={onClickMock} />
    );
    screen.getByText("Mark as Finished");
    expect(onClickMock).toHaveBeenCalledTimes(1);
});

it('Fire event button click changes status from Read Again to Mark as Finished', () => {
    const { rerender } = render(<BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={true} onClick={onClickMock} />);
    screen.getByText("Read Again");
    fireEvent.click(screen.getByText('Read Again'));
    rerender(
        <BookCard isInLibrary={isInLibraryMock.mockReturnValue(true)} isFinished={false} onClick={onClickMock} />
    );
    screen.getByText("Mark as Finished");
    expect(onClickMock).toHaveBeenCalledTimes(1);
});