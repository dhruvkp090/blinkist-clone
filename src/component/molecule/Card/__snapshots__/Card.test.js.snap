// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot with default image and finished book 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <img
      alt="Animal Farm"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://images.blinkist.com/images/books/534d4cb46633610007320000/1_1/640.jpg"
      title="Animal Farm"
    />
    <div
      class="MuiCardContent-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
      >
        Animal Farm
      </h6>
      <p
        class="MuiTypography-root author MuiTypography-body1 MuiTypography-colorSecondary MuiTypography-gutterBottom"
      >
        <strong>
          Gorge Orwell
        </strong>
      </p>
      <div
        class="timeSection"
      >
        <img
          alt=""
          class="icon"
          src="https://d17pjsg7x52x9r.cloudfront.net/assets/components/book_card/clock-b0e2e0235fbe1df824d662b2b3b96611e3711bf5b5c7556b8bd3828720f86dbc.svg"
        />
        <p
          class="MuiTypography-root timeText MuiTypography-body2 MuiTypography-gutterBottom"
        >
          20-minute read
        </p>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-fullWidth"
        role="button"
        style="justify-content: center;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            <strong>
              Read Again
            </strong>
          </p>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshot with passed image and unfinished book 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <img
      alt="Animal Farm"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://images.blinkist.com/images/books/608aa9b16cee070007228a70/1_1/640.jpg"
      title="Animal Farm"
    />
    <div
      class="MuiCardContent-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
      >
        Animal Farm
      </h6>
      <p
        class="MuiTypography-root author MuiTypography-body1 MuiTypography-colorSecondary MuiTypography-gutterBottom"
      >
        <strong>
          Gorge Orwell
        </strong>
      </p>
      <div
        class="timeSection"
      >
        <img
          alt=""
          class="icon"
          src="https://d17pjsg7x52x9r.cloudfront.net/assets/components/book_card/clock-b0e2e0235fbe1df824d662b2b3b96611e3711bf5b5c7556b8bd3828720f86dbc.svg"
        />
        <p
          class="MuiTypography-root timeText MuiTypography-body2 MuiTypography-gutterBottom"
        >
          20-minute read
        </p>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        role="button"
        style="justify-content: center;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            <strong>
              Mark as Finished
            </strong>
          </p>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
